File: Minimal/tests/Brontide.Minimal.ComponentManagement.Tests/FixtureTests.fs
namespace Brontide.Minimal.ComponentManagement.Tests

open System
open System.IO
open NUnit.Framework
open Brontide.Minimal.Experimental.ComponentManagement

[<TestFixture>]
type FixtureTests() =

    let fixturePath name =
        Path.Combine(TestContext.CurrentContext.TestDirectory, "component-management", "fixtures", name)

    let catalogJson () = File.ReadAllText(fixturePath "cm0-catalog.json")
    let miceJson () = File.ReadAllText(fixturePath "cm0-mice-topology.json")

    /// Returns true when the thunk raises ArgumentException; keeps NUnit delegate overloads out of F#.
    let raisesArgument (action: unit -> unit) : bool =
        try
            action ()
            false
        with :? ArgumentException -> true

    /// Runs a loader expected to fail closed and returns its accumulated failure list.
    let rejection (load: unit -> unit) : string list =
        try
            load ()
            Assert.Fail "expected FixtureFormatException but loading succeeded"
            []
        with FixtureFormatException failures -> failures

    /// Structural equality assertion; sidesteps NUnit's ambiguous list/nullable EqualTo overloads.
    let equal (actual: 'T) (expected: 'T) =
        Assert.That((actual = expected), Is.True, sprintf "expected %A but was %A" expected actual)

    [<TestCase("")>]
    [<TestCase("Has Spaces")>]
    [<TestCase("UPPER")>]
    [<TestCase("under_score")>]
    member _.``Identifier creation rejects invalid syntax``(value: string) =
        Assert.That(raisesArgument (fun () -> SourceId.create value |> ignore), Is.True)

    [<Test>]
    member _.``Identifier spaces are distinct types over one primitive``() =
        let source = SourceId.create "shared.token"
        let publisher = PublisherId.create "shared.token"
        equal (SourceId.value source) (PublisherId.value publisher)
        Assert.That((source.GetType() <> publisher.GetType()), Is.True)

    [<TestCase("1..1", 1)>]
    [<TestCase("0..*", 0)>]
    member _.``Cardinality parses declared forms``(text: string, minimum: int) =
        equal (Cardinality.parse text).Minimum minimum

    [<TestCase("1")>]
    [<TestCase("2..1")>]
    [<TestCase("-1..2")>]
    member _.``Cardinality rejects invalid forms``(text: string) =
        Assert.That(raisesArgument (fun () -> Cardinality.parse text |> ignore), Is.True)

    [<Test>]
    member _.``Catalog fixture loads with expected shape``() =
        let fixture = FixtureLoader.loadCatalog (catalogJson ())
        equal (List.length fixture.Sources) 3
        equal (List.length fixture.Packages) 5
        equal (List.length fixture.ComponentDefinitions) 5
        equal (List.length fixture.ActivatedOccurrences) 3
        equal (List.length fixture.Storefront) 3

    [<Test>]
    member _.``Catalog expectations surface every required cm0 case``() =
        let expectations = (FixtureLoader.loadCatalog (catalogJson ())).Expectations
        equal expectations.DuplicateComponentIdentityAcrossSources [ PackageId.create "pkg.contoso.cooling" ]
        equal expectations.MirroredPublishers [ PublisherId.create "pub.contoso" ]
        equal expectations.MultiPublisherSources [ SourceId.create "src.bazaar" ]
        equal (List.length expectations.ContractsWithSeveralDefinitions) 2
        equal expectations.DefinitionsWithSeveralOccurrences [ DefinitionId.create "def.northwind.telemetry" ]
        equal (List.length expectations.OccupiedBindings) 2
        equal expectations.SystemDefaultScopes [ BindingScopeId.create "scope.system" ]
        equal (List.length expectations.ExplicitPreferences) 1
        equal expectations.GenericCandidates [ DefinitionId.create "def.contoso.generic-telemetry" ]
        equal expectations.ConflictingVersionClaims [ PackageId.create "pkg.contoso.cooling" ]
        equal expectations.MissingArtifacts [ ArtifactId.create "art.missing-db" ]
        equal (List.length expectations.ContradictoryEvidence) 1

    [<Test>]
    member _.``Catalog loading is deterministic across repeated loads``() =
        let first = FixtureLoader.loadCatalog (catalogJson ())
        let second = FixtureLoader.loadCatalog (catalogJson ())
        equal second.Packages first.Packages
        equal second.Advertisements first.Advertisements
        equal second.OccupiedBindings first.OccupiedBindings
        equal second.Expectations.MissingArtifacts first.Expectations.MissingArtifacts

    [<Test>]
    member _.``Loading grants nothing and preserves claim observation distinction``() =
        let fixture = FixtureLoader.loadCatalog (catalogJson ())
        let advertised =
            fixture.Advertisements
            |> List.find (fun a -> a.Source = SourceId.create "src.bazaar" && a.Package = PackageId.create "pkg.contoso.cooling")
        let declared = fixture.Packages |> List.find (fun p -> p.Package = PackageId.create "pkg.contoso.cooling")
        Assert.That(advertised.AdvertisedVersion, Is.Not.EqualTo declared.Version)

    [<Test>]
    member _.``Malformed json is rejected``() =
        let failures = rejection (fun () -> FixtureLoader.loadCatalog "{ not json" |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "not valid JSON"), Is.True)

    [<Test>]
    member _.``Unknown schema version is rejected``() =
        let mutated = (catalogJson ()).Replace("\"schemaVersion\": 1", "\"schemaVersion\": 2")
        let failures = rejection (fun () -> FixtureLoader.loadCatalog mutated |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "schemaVersion"), Is.True)

    [<Test>]
    member _.``Duplicate identifier is rejected by name``() =
        let mutated = (catalogJson ()).Replace("\"source\": \"src.contoso-mirror\"", "\"source\": \"src.local-cache\"")
        let failures = rejection (fun () -> FixtureLoader.loadCatalog mutated |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "duplicate identifier 'src.local-cache'"), Is.True)

    [<Test>]
    member _.``Unresolved reference is rejected by name``() =
        let mutated = (catalogJson ()).Replace("\"package\": \"pkg.fabrikam.cooling\", \"advertisedVersion\"", "\"package\": \"pkg.unknown\", \"advertisedVersion\"")
        let failures = rejection (fun () -> FixtureLoader.loadCatalog mutated |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "pkg.unknown"), Is.True)

    [<Test>]
    member _.``Unknown top level property is rejected``() =
        let mutated = (catalogJson ()).Replace("\"schemaVersion\": 1,", "\"schemaVersion\": 1, \"surprise\": true,")
        let failures = rejection (fun () -> FixtureLoader.loadCatalog mutated |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "unknown property 'surprise'"), Is.True)

    [<Test>]
    member _.``Undeclared missing artifact is rejected``() =
        let mutated = (catalogJson ()).Replace("\"missingArtifacts\": [\"art.missing-db\"]", "\"missingArtifacts\": []")
        let failures = rejection (fun () -> FixtureLoader.loadCatalog mutated |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "art.missing-db"), Is.True)

    [<Test>]
    member _.``Digest mismatch is rejected``() =
        let mutated = (catalogJson ()).Replace("fake-artifact:contoso-cooling:1.4.0", "fake-artifact:tampered")
        let failures = rejection (fun () -> FixtureLoader.loadCatalog mutated |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "digest mismatch"), Is.True)

    [<Test>]
    member _.``Expectation disagreeing with data is rejected``() =
        let mutated = (catalogJson ()).Replace("\"genericCandidates\": [\"def.contoso.generic-telemetry\"]", "\"genericCandidates\": []")
        let failures = rejection (fun () -> FixtureLoader.loadCatalog mutated |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "genericCandidates"), Is.True)

    [<Test>]
    member _.``Mice fixture keeps two distinct nodes with four functions each``() =
        let fixture = FixtureLoader.loadMiceTopology (miceJson ())
        let nodesWithFunctions = fixture.Functions |> List.map (fun f -> f.Node) |> List.distinct
        equal (List.length nodesWithFunctions) 2
        equal fixture.Expectations.FunctionsPerMouseNode 4
        equal (fixture.Functions |> List.filter (fun f -> f.Node = TopologyNodeId.create "node.mouse-a") |> List.length) 4
        equal (fixture.Functions |> List.filter (fun f -> f.Node = TopologyNodeId.create "node.mouse-b") |> List.length) 4

    [<Test>]
    member _.``Mice fixture surfaces malicious and contradictory claims without dropping them``() =
        let fixture = FixtureLoader.loadMiceTopology (miceJson ())
        equal fixture.Expectations.MaliciousClaims [ ClaimId.create "claim.b-hosts-root"; ClaimId.create "claim.b-owns-a" ]
        equal (List.length fixture.Expectations.ContradictoryClaims) 1
        Assert.That(fixture.Claims |> List.map (fun c -> c.Claim) |> List.contains (ClaimId.create "claim.b-owns-a"), Is.True)

    [<Test>]
    member _.``Mice fixture rejects unclassified claims``() =
        let mutated = (miceJson ()).Replace("\"maliciousClaims\": [\"claim.b-hosts-root\", \"claim.b-owns-a\"]", "\"maliciousClaims\": [\"claim.b-owns-a\"]")
        let failures = rejection (fun () -> FixtureLoader.loadMiceTopology mutated |> ignore)
        Assert.That(failures |> List.exists (fun f -> f.Contains "claim.b-hosts-root"), Is.True)

File: Reference/tests/Brontide.Reference.ComponentManagement.Tests/FixtureTests.cs
using Brontide.Reference.Experimental.ComponentManagement;
using NUnit.Framework;

namespace Brontide.Reference.ComponentManagement.Tests;

[TestFixture]
public sealed class FixtureTests
{
    private static string FixturePath(string name) =>
        Path.Combine(TestContext.CurrentContext.TestDirectory, "component-management", "fixtures", name);

    private static string CatalogJson() => File.ReadAllText(FixturePath("cm0-catalog.json"));

    private static string MiceJson() => File.ReadAllText(FixturePath("cm0-mice-topology.json"));

    [TestCase("")]
    [TestCase("Has Spaces")]
    [TestCase("UPPER")]
    [TestCase("under_score")]
    public void Identifier_creation_rejects_invalid_syntax(string value)
    {
        Assert.That(() => SourceId.Create(value), Throws.ArgumentException);
    }

    [Test]
    public void Identifier_spaces_are_distinct_types_over_one_primitive()
    {
        var source = SourceId.Create("shared.token");
        var publisher = PublisherId.Create("shared.token");
        Assert.Multiple(() =>
        {
            Assert.That(source.Value, Is.EqualTo(publisher.Value));
            Assert.That(source.GetType(), Is.Not.EqualTo(publisher.GetType()));
        });
    }

    [TestCase("1..1", 1, 1)]
    [TestCase("0..*", 0, null)]
    public void Cardinality_parses_declared_forms(string text, int minimum, int? maximum)
    {
        var cardinality = Cardinality.Parse(text);
        Assert.Multiple(() =>
        {
            Assert.That(cardinality.Minimum, Is.EqualTo(minimum));
            Assert.That(cardinality.Maximum, Is.EqualTo(maximum));
        });
    }

    [TestCase("1")]
    [TestCase("2..1")]
    [TestCase("-1..2")]
    public void Cardinality_rejects_invalid_forms(string text)
    {
        Assert.That(() => Cardinality.Parse(text), Throws.ArgumentException);
    }

    [Test]
    public void Catalog_fixture_loads_with_expected_shape()
    {
        var fixture = FixtureLoader.LoadCatalog(CatalogJson());
        Assert.Multiple(() =>
        {
            Assert.That(fixture.Sources, Has.Count.EqualTo(3));
            Assert.That(fixture.Packages, Has.Count.EqualTo(5));
            Assert.That(fixture.ComponentDefinitions, Has.Count.EqualTo(5));
            Assert.That(fixture.ActivatedOccurrences, Has.Count.EqualTo(3));
            Assert.That(fixture.Storefront, Has.Count.EqualTo(3));
        });
    }

    [Test]
    public void Catalog_expectations_surface_every_required_cm0_case()
    {
        var expectations = FixtureLoader.LoadCatalog(CatalogJson()).Expectations;
        Assert.Multiple(() =>
        {
            Assert.That(expectations.DuplicateComponentIdentityAcrossSources, Is.EqualTo(new[] { PackageId.Create("pkg.contoso.cooling") }));
            Assert.That(expectations.MirroredPublishers, Is.EqualTo(new[] { PublisherId.Create("pub.contoso") }));
            Assert.That(expectations.MultiPublisherSources, Is.EqualTo(new[] { SourceId.Create("src.bazaar") }));
            Assert.That(expectations.ContractsWithSeveralDefinitions, Has.Count.EqualTo(2));
            Assert.That(expectations.DefinitionsWithSeveralOccurrences, Is.EqualTo(new[] { DefinitionId.Create("def.northwind.telemetry") }));
            Assert.That(expectations.OccupiedBindings, Has.Count.EqualTo(2));
            Assert.That(expectations.SystemDefaultScopes, Is.EqualTo(new[] { BindingScopeId.Create("scope.system") }));
            Assert.That(expectations.ExplicitPreferences, Has.Count.EqualTo(1));
            Assert.That(expectations.GenericCandidates, Is.EqualTo(new[] { DefinitionId.Create("def.contoso.generic-telemetry") }));
            Assert.That(expectations.ConflictingVersionClaims, Is.EqualTo(new[] { PackageId.Create("pkg.contoso.cooling") }));
            Assert.That(expectations.MissingArtifacts, Is.EqualTo(new[] { ArtifactId.Create("art.missing-db") }));
            Assert.That(expectations.ContradictoryEvidence, Has.Count.EqualTo(1));
        });
    }

    [Test]
    public void Catalog_loading_is_deterministic_across_repeated_loads()
    {
        var first = FixtureLoader.LoadCatalog(CatalogJson());
        var second = FixtureLoader.LoadCatalog(CatalogJson());
        Assert.Multiple(() =>
        {
            Assert.That(second.Packages, Is.EqualTo(first.Packages));
            Assert.That(second.Advertisements, Is.EqualTo(first.Advertisements));
            Assert.That(second.OccupiedBindings, Is.EqualTo(first.OccupiedBindings));
            Assert.That(second.Expectations.MissingArtifacts, Is.EqualTo(first.Expectations.MissingArtifacts));
        });
    }

    [Test]
    public void Loading_grants_nothing_and_preserves_claim_observation_distinction()
    {
        var fixture = FixtureLoader.LoadCatalog(CatalogJson());
        var advertised = fixture.Advertisements.Single(a =>
            a.Source == SourceId.Create("src.bazaar") && a.Package == PackageId.Create("pkg.contoso.cooling"));
        var declared = fixture.Packages.Single(p => p.Package == PackageId.Create("pkg.contoso.cooling"));
        Assert.That(advertised.AdvertisedVersion, Is.Not.EqualTo(declared.Version),
            "the source's claim must stay distinguishable from the package's declared version");
    }

    [Test]
    public void Malformed_json_is_rejected()
    {
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadCatalog("{ not json"));
        Assert.That(exception!.Failures, Has.Some.Contains("not valid JSON"));
    }

    [Test]
    public void Unknown_schema_version_is_rejected()
    {
        var mutated = CatalogJson().Replace("\"schemaVersion\": 1", "\"schemaVersion\": 2", StringComparison.Ordinal);
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadCatalog(mutated));
        Assert.That(exception!.Failures, Has.Some.Contains("schemaVersion"));
    }

    [Test]
    public void Duplicate_identifier_is_rejected_by_name()
    {
        var mutated = CatalogJson().Replace("\"source\": \"src.contoso-mirror\"", "\"source\": \"src.local-cache\"", StringComparison.Ordinal);
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadCatalog(mutated));
        Assert.That(exception!.Failures, Has.Some.Contains("duplicate identifier 'src.local-cache'"));
    }

    [Test]
    public void Unresolved_reference_is_rejected_by_name()
    {
        var mutated = CatalogJson().Replace("\"package\": \"pkg.fabrikam.cooling\", \"advertisedVersion\"", "\"package\": \"pkg.unknown\", \"advertisedVersion\"", StringComparison.Ordinal);
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadCatalog(mutated));
        Assert.That(exception!.Failures, Has.Some.Contains("pkg.unknown"));
    }

    [Test]
    public void Unknown_top_level_property_is_rejected()
    {
        var mutated = CatalogJson().Replace("\"schemaVersion\": 1,", "\"schemaVersion\": 1, \"surprise\": true,", StringComparison.Ordinal);
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadCatalog(mutated));
        Assert.That(exception!.Failures, Has.Some.Contains("unknown property 'surprise'"));
    }

    [Test]
    public void Undeclared_missing_artifact_is_rejected()
    {
        var mutated = CatalogJson().Replace("\"missingArtifacts\": [\"art.missing-db\"]", "\"missingArtifacts\": []", StringComparison.Ordinal);
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadCatalog(mutated));
        Assert.That(exception!.Failures, Has.Some.Contains("art.missing-db"));
    }

    [Test]
    public void Digest_mismatch_is_rejected()
    {
        var mutated = CatalogJson().Replace("fake-artifact:contoso-cooling:1.4.0", "fake-artifact:tampered", StringComparison.Ordinal);
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadCatalog(mutated));
        Assert.That(exception!.Failures, Has.Some.Contains("digest mismatch"));
    }

    [Test]
    public void Expectation_disagreeing_with_data_is_rejected()
    {
        var mutated = CatalogJson().Replace("\"genericCandidates\": [\"def.contoso.generic-telemetry\"]", "\"genericCandidates\": []", StringComparison.Ordinal);
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadCatalog(mutated));
        Assert.That(exception!.Failures, Has.Some.Contains("genericCandidates"));
    }

    [Test]
    public void Mice_fixture_keeps_two_distinct_nodes_with_four_functions_each()
    {
        var fixture = FixtureLoader.LoadMiceTopology(MiceJson());
        var nodesWithFunctions = fixture.Functions.Select(f => f.Node).Distinct().ToArray();
        Assert.Multiple(() =>
        {
            Assert.That(nodesWithFunctions, Has.Length.EqualTo(2));
            Assert.That(fixture.Expectations.FunctionsPerMouseNode, Is.EqualTo(4));
            Assert.That(fixture.Functions.Count(f => f.Node == TopologyNodeId.Create("node.mouse-a")), Is.EqualTo(4));
            Assert.That(fixture.Functions.Count(f => f.Node == TopologyNodeId.Create("node.mouse-b")), Is.EqualTo(4));
        });
    }

    [Test]
    public void Mice_fixture_surfaces_malicious_and_contradictory_claims_without_dropping_them()
    {
        var fixture = FixtureLoader.LoadMiceTopology(MiceJson());
        Assert.Multiple(() =>
        {
            Assert.That(fixture.Expectations.MaliciousClaims, Is.EqualTo(new[]
            {
                ClaimId.Create("claim.b-hosts-root"),
                ClaimId.Create("claim.b-owns-a"),
            }));
            Assert.That(fixture.Expectations.ContradictoryClaims, Has.Count.EqualTo(1));
            Assert.That(fixture.Claims.Select(c => c.Claim), Does.Contain(ClaimId.Create("claim.b-owns-a")),
                "malicious claims stay recorded as attributable data rather than being silently dropped");
        });
    }

    [Test]
    public void Mice_fixture_rejects_unclassified_claims()
    {
        var mutated = MiceJson().Replace(
            "\"maliciousClaims\": [\"claim.b-hosts-root\", \"claim.b-owns-a\"]",
            "\"maliciousClaims\": [\"claim.b-owns-a\"]",
            StringComparison.Ordinal);
        var exception = Assert.Throws<FixtureFormatException>(() => FixtureLoader.LoadMiceTopology(mutated));
        Assert.That(exception!.Failures, Has.Some.Contains("claim.b-hosts-root"));
    }
}
